app/models/upload.rb

```diff
@@ -316,6 +317,64 @@ def thumbnail_height
     get_dimension(:thumbnail_height)
   end
 
+  def dominant_color(calculate_if_missing: false)
+    val = read_attribute(:dominant_color)
+    if val.nil? && calculate_if_missing
+      calculate_dominant_color!
+      read_attribute(:dominant_color)
+    else
+      val
+    end
+  end
+
+  def calculate_dominant_color!(local_path = nil)
+    color = nil
+
+    if !FileHelper.is_supported_image?("image.#{extension}") || extension == "svg"
+      color = ""
+    end
+
+    if color.nil?
+      local_path ||=
+        if local?
+          Discourse.store.path_for(self)
+        else
+          Discourse.store.download(self).path
+        end
+
+      color = begin
+        data = Discourse::Utils.execute_command(
+          "nice",
+          "-n",
+          "10",
+          "convert",
+          local_path,
+          "-colors",
+          "1",
+          "-define",
+          "histogram:unique-colors=true",
+          "-format",
+          "%c",
+          "histogram:info:",
+          timeout: MAX_DOMINANT_COLOR_SECONDS
+        )
+
+        color = data[/#([0-9A-F]{6})/, 1]
+        raise "Unable to parse dominant color output:\n#{data}" if color.nil?
```

Comment:
I wonder if we need to differentiate between unable to parse because the regex doesn't work and invalid `data` because of timeout etc...

Reply:
:+1: updated. non-zero imagemagick exits and timeouts will be swallowed silently (they can happen due to bad end-user input, so no need to create log noise).

String parsing errors will raise true exceptions - this should never happen, so it should not be swallowed silently